js/blockbid.js

```diff
@@ -0,0 +1,657 @@
+        for (let i = 0; i < response['asks'].length; i++) {
+            arrAsks.push ([response['asks'][i]['price'], response['asks'][i]['volume'], []]);
+        }
+        preParseBook['bids'] = arrBids;
+        preParseBook['asks'] = arrAsks;
+        return this.parseOrderBook (preParseBook, undefined, 'bids', 'asks', 0, 1);
+    }
+
+    parseTrade (trade, market = undefined) {
+        let symbol = undefined;
+        if (market) {
+            symbol = market['symbol'];
+        }
+        let datetime = trade['createdAt'];
+        let timestamp = this.parse8601 (datetime);
+        let price = this.safeFloat (trade, 'price');
+        let amount = this.safeFloat (trade, 'volume');
+        let cost = price * amount;
+        return {
+            'info': trade,
+            'timestamp': timestamp,
+            'datetime': datetime,
+            'symbol': symbol,
+            'id': trade['id'],
+            'order': undefined,
+            'type': undefined,
+            'side': trade['side'],
+            'price': price,
+            'amount': amount,
+            'cost': cost,
+            'fee': undefined,
+        };
+    }
+
+    async fetchTrades (symbol, since = undefined, limit = 50, params = {}) {
+        await this.loadMarkets ();
+        let market = this.market (symbol);
+        let response = await this.publicGetTrades (this.extend ({
+            'market': market['id'],
+            'limit': limit,
+        }, params));
+        let err = this.handleError (response);
+        if (err) {
+            throw new ExchangeError (this.id + ' has thrown an error: ' + err);
+        }
+        return this.parseTrades (response, market, since, limit);
+    }
+
+    parseOHLCV (ohlcv, market = undefined, timeframe = '5m', since = undefined, limit = undefined) {
+        let datetime = this.parse8601 (ohlcv['timestamp']);
+        return [
+            datetime,
+            parseFloat (ohlcv['open']),
+            parseFloat (ohlcv['high']),
+            parseFloat (ohlcv['low']),
+            parseFloat (ohlcv['close']),
+            parseFloat (ohlcv['volume']),
+        ];
+    }
+
+    async fetchOHLCV (symbol, timeframe = '1m', since = undefined, limit = undefined, params = {}) {
+        await this.loadMarkets ();
+        let market = this.market (symbol);
+        let request = {
+            'market': market['id'],
+            'period': this.timeframes[timeframe],
+        };
+        if (typeof since !== 'undefined') {
+            request['timestamp'] = since;
+        }
+        let response = await this.publicGetOhlc (this.extend (request, params));
+        let err = this.handleError (response);
+        if (err) {
+            throw new ExchangeError (this.id + ' has thrown an error: ' + err);
+        }
+        return this.parseOHLCVs (response, market, timeframe, since, limit);
+    }
+
+    async fetchBalances (params = {}) {
+        await this.loadMarkets ();
+        if (!this.apiKey || !this.secret) {
```

Comment:
We don't want to check for keys in every private method, instead we have a special exchange property `requiredCredentials` which looks like so: https://github.com/ccxt/ccxt/blob/master/js/base/Exchange.js#L138 and can be overrided for exchange specific needs like so: https://github.com/ccxt/ccxt/blob/master/js/bitstamp.js#L34 and those required credentials are checked once in the `sign` method (in the private codepath), like so: https://github.com/ccxt/ccxt/blob/master/js/bitstamp.js#L854